tests/formats/cml.cpp

```diff
@@ -0,0 +1,198 @@
+        CHECK(cell.b() == 3.0);
+        CHECK(cell.c() == 3.0);
+        CHECK(cell.alpha() == 90.0);
+        CHECK(cell.beta() == 90.0);
+        CHECK(cell.gamma() == 120.0);
+
+        auto positions = frame.positions();
+        auto fract0 = frame.cell().matrix().invert() * positions[0];
+        CHECK(approx_eq(fract0, Vector3D(-1.77493, 0.980333, 0.0000), 1e-3));
+    }
+
+    SECTION("Read properties") {
+        Trajectory file("data/cml/properties.cml");
+        auto frame = file.read();
+        CHECK(frame.get("num_c")->as_double() == 10.0);
+        CHECK(frame.get("is_organic")->as_bool() == true);
+        CHECK(frame[6].get("r")->as_string() == "tButyl");
+    }
+}
+
+TEST_CASE("Write CML file") {
+    auto tmpfile = NamedTempPath(".cml");
+    auto frame = Frame();
+    frame.add_atom(Atom("A"), {1, 2, 3});
+    frame.add_atom(Atom("B"), {1, 2, 3});
+    frame.add_atom(Atom("C"), {1, 2, 3});
+    frame.add_atom(Atom("D"), {1, 2, 3});
+
+    auto file = Trajectory(tmpfile, 'w');
+    file.write(frame);
+
+    frame.positions()[0] = {4.0, 5.0, 6.0};
+    frame.set_cell(UnitCell(22));
+
+    frame.set("is_organic", false);
+    frame.set("name", "test");
+
+    frame[1].set_mass(12);
+
+    frame[2].set("force", Vector3D{1., 2., 3.});
+    frame[2].set("num_c", 1.0);
+
+    frame[3].set_charge(1.0);
+    frame[3].set("hydrogen_count", 3);
+
+    frame.add_bond(0, 1, Bond::UNKNOWN);
+    frame.add_bond(0, 2, Bond::SINGLE);
+    frame.add_bond(1, 2, Bond::DOUBLE);
+    frame.add_bond(1, 3, Bond::TRIPLE);
+    frame.add_bond(2, 3, Bond::AROMATIC);
+    
+    file.write(frame);
+
+    file.close();
+
+    // We can't compare the files directly as the properties may be in any order
+    // because property_map is based on an unordered_map. Therefore, we should
+    // just try to reload the file and see if everything is as it should be.
+
+    auto check_cml = Trajectory(tmpfile);
+    CHECK(check_cml.nsteps() == 2);
+    
+    auto frame1 = check_cml.read();
+    CHECK(frame1.size() == 4);
+    CHECK(frame1.topology().bonds().size() == 0);
+
+    auto frame2 = check_cml.read();
+    CHECK(frame2.size() == 4);
+    CHECK(frame2.cell() == UnitCell(22));
+
+    auto& orders = frame2.topology().bond_orders();
+    CHECK(orders[0] == Bond::UNKNOWN);
+    CHECK(orders[1] == Bond::SINGLE);
+    CHECK(orders[2] == Bond::DOUBLE);
+    CHECK(orders[3] == Bond::TRIPLE);
+    CHECK(orders[4] == Bond::AROMATIC);
+
+    CHECK(frame2.get("is_organic")->as_bool() == false);
+    CHECK(frame2.get("name")->as_string() == "test");
+
+    CHECK(frame2[1].mass() == 12);
+
+    CHECK(frame2[2].get("num_c")->as_double() == 1.0);
+    CHECK(frame2[2].get("force")->as_vector3d() == Vector3D{1., 2., 3.});
+
+    CHECK(frame2[3].charge() == 1);
+    CHECK(frame2[3].get("hydrogen_count")->as_double() == 3);
+}
+
+TEST_CASE("Append CML file") {
+    auto tmpfile = NamedTempPath(".cml");
+    const auto EXPECTED_CONTENT =
+    "<molecule title=\"appended\">\n"
```

Comment:
You can use C++ raw strings here to remove most of the `\n` and `\"`:

```cpp
auto string = R"(foo bar "yeah"

no explicit new line \o/)";
```

Reply:
I think a lot of other formats need this done as well! Maybe I'll make another PR for that.